lib/pos_epoch.go

```diff
@@ -0,0 +1,171 @@
+	}
+	if epochEntry != nil {
+		// Cache in the UtxoView.
+		bav.CurrentEpochEntry = epochEntry.Copy()
+	}
+	return epochEntry, nil
+}
+
+func (bav *UtxoView) GetCurrentEpochNumber() (uint64, error) {
+	epochEntry, err := bav.GetCurrentEpochEntry()
+	if err != nil {
+		return 0, errors.Wrapf(err, "UtxoView.GetCurrentEpochNumber: ")
+	}
+	if epochEntry == nil {
+		return 0, errors.New("UtxoView.GetCurrentEpochNumber: no CurrentEpochEntry found")
+	}
+	return epochEntry.EpochNumber, nil
+}
+
+func (bav *UtxoView) _setCurrentEpochEntry(epochEntry *EpochEntry) {
+	if epochEntry == nil {
+		glog.Errorf("UtxoView._setCurrentEpochEntry: called with nil EpochEntry")
+		return
+	}
+	bav.CurrentEpochEntry = epochEntry.Copy()
+}
+
+func (bav *UtxoView) _flushCurrentEpochEntryToDbWithTxn(txn *badger.Txn, blockHeight uint64) error {
+	if bav.CurrentEpochEntry == nil {
+		// It is possible that the current UtxoView never interacted with the CurrentEpochEntry
+		// in which case the CurrentEpochEntry in the UtxoView will be nil. In that case, we
+		// don't want to overwrite what is in the database. Just no-op.
+		return nil
+	}
+	if err := DBPutCurrentEpochEntryWithTxn(txn, bav.Snapshot, bav.CurrentEpochEntry, blockHeight); err != nil {
+		return errors.Wrapf(err, "_flushCurrentEpochEntryToDbWithTxn: ")
+	}
+	return nil
+}
+
+//
+// DB UTILS
+//
+
+func DBKeyForCurrentEpoch() []byte {
+	return append([]byte{}, Prefixes.PrefixCurrentEpoch...)
+}
+
+func DBGetCurrentEpochEntry(handle *badger.DB, snap *Snapshot) (*EpochEntry, error) {
+	var ret *EpochEntry
+	err := handle.View(func(txn *badger.Txn) error {
+		var innerErr error
+		ret, innerErr = DBGetCurrentEpochEntryWithTxn(txn, snap)
+		return innerErr
+	})
+	return ret, err
+}
+
+func DBGetCurrentEpochEntryWithTxn(txn *badger.Txn, snap *Snapshot) (*EpochEntry, error) {
+	// Retrieve StakeEntry from db.
+	key := DBKeyForCurrentEpoch()
+	epochEntryBytes, err := DBGetWithTxn(txn, snap, key)
+	if err != nil {
+		// We don't want to error if the key isn't found. Instead, return nil.
+		if err == badger.ErrKeyNotFound {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "DBGetCurrentEpochEntry: problem retrieving EpochEntry: ")
+	}
+
+	// Decode EpochEntry from bytes.
+	rr := bytes.NewReader(epochEntryBytes)
+	epochEntry, err := DecodeDeSoEncoder(&EpochEntry{}, rr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "DBGetCurrentEpochEntry: problem decoding EpochEntry: ")
+	}
+	return epochEntry, nil
+}
+
+func DBPutCurrentEpochEntryWithTxn(txn *badger.Txn, snap *Snapshot, epochEntry *EpochEntry, blockHeight uint64) error {
+	// Set EpochEntry in PrefixCurrentEpoch.
+	if epochEntry == nil {
+		// This is just a safety check that we are not accidentally overwriting an
+		// existing EpochEntry with a nil EpochEntry. This should never happen.
+		glog.Errorf("DBPutCurrentEpochEntryWithTxn: called with nil EpochEntry")
+		return nil
```

Comment:
since this function does return an error, should we just return the error instead of glogging?